### PR TITLE
[chat] fix value computing bug

### LINKED ISSUE #3622: [BUG]: [chat] Using wrong part of values to compute value
### 🐛 Describe the bug

[https://github.com/hpcaitech/ColossalAI/blob/b0ce5a10326912961f0bc07cbbd250bab7b9c399/applications/Chat/coati/models/base/critic.py#L45-L50](https://github.com/hpcaitech/ColossalAI/blob/b0ce5a10326912961f0bc07cbbd250bab7b9c399/applications/Chat/coati/models/base/critic.py#L45-L50)

We should use the generation part of the sequenses to compute value. So the indexing should be `[-num_actions:]` instead of `[:-num_actions]` and using `action_mask`.

### Environment

_No response_

### PR DESCRIPTION
**Pull request by zhang-yi-chi**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

fixed #3622 

## 📝 What does this PR do?

[https://github.com/hpcaitech/ColossalAI/blob/b0ce5a10326912961f0bc07cbbd250bab7b9c399/applications/Chat/coati/models/base/critic.py#L45-L50](https://github.com/hpcaitech/ColossalAI/blob/b0ce5a10326912961f0bc07cbbd250bab7b9c399/applications/Chat/coati/models/base/critic.py#L45-L50)

We should use the generation part of the sequenses to compute value. So the indexing should be `[-num_actions:]` instead of `[:-num_actions]` and using `action_mask`.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: applications/Chat/coati/experience_maker/naive.py [hunk 1]
```diff
@@ -24,7 +24,7 @@ def make_experience(self, input_ids: torch.Tensor, **generate_kwargs) -> Experie
         action_log_probs = self.actor(sequences, num_actions, attention_mask)
         base_action_log_probs = self.initial_model(sequences, num_actions, attention_mask)
         value = self.critic(sequences, action_mask, attention_mask)
-        r = self.reward_model(sequences, attention_mask)
+        r = self.reward_model(sequences, attention_mask, action_mask=action_mask)
```
Comment (ht-zhou) [inline applications/Chat/coati/experience_maker/naive.py:27]: To compute reward，we should use the whole sequence as training rm

File: applications/Chat/coati/models/base/critic.py [hunk 2]
```diff
@@ -44,9 +44,8 @@ def forward(self,
 
         if action_mask is not None and self.use_action_mask:
```
Comment (ht-zhou) [inline applications/Chat/coati/models/base/critic.py:45]: Using action mask here means to only use state (prompt) to compute value, as described in RL patterns.

**Comment by ht-zhou** [comment]
Also we are updating our PPO pattern recently, the modeling of MDP will be changed soon.

File: applications/Chat/coati/models/base/critic.py (hunk 2 above)
Comment (zhang-yi-chi) [reply applications/Chat/coati/models/base/critic.py:45]: Sorry for my misunderstanding. So here uses one step MDP if I understand this correctly? Maybe we should change to multi step MDP and compute values for every answer input.

File: applications/Chat/coati/experience_maker/naive.py (hunk 1 above)
Comment (zhang-yi-chi) [reply applications/Chat/coati/experience_maker/naive.py:27]: Yes you are right, here we should follow the rm training. 

This remind me of another issue: either log-sig or log-exp computes (chosen_reward - reject_reward) which actually remove the prompt's reward if we ignore the average computation.  Does this means the prompt part of the sequence are under Inadequate training?

**Comment by zhang-yi-chi** [comment]
> Also we are updating our PPO pattern recently, the modeling of MDP will be changed soon.

Looking forward to seeing the changes. Closed this pull request.

**Comment by ht-zhou** [comment]
Anyway, thanks for your contribution!
